src/llvmir2hll/graphs/cfg/cfg_traversal.cpp

```diff
@@ -163,12 +176,78 @@ bool CFGTraversal::getCurrRetVal() const {
 * @param[in] node Node to be traversed.
 * @param[in] stmtIter Iterator to a statement in @a node to be checked.
 *
-* If @a stmtIter equals @c node->stmt_end(), the function traverses all
-* successors of @a node.
+* This function traverses the entire CFG rooted at @a node, starting with the statement
+* in @a stmtIter.
 */
 bool CFGTraversal::performTraversalImpl(ShPtr<CFG::Node> node,
 		CFG::stmt_iterator stmtIter) {
-	while (stmtIter != node->stmt_end()) {
+	// The stack nodes consist of the current CFG node, a statement iterator storing the current
+	// place in the node's statements, and an iterator storing the current place in the node's
+	// successor list.
+	// Statement place is only needed because we may be called with a statement iterator
+	// that doesn't point to the beginning.
+	using NodeType = std::tuple<ShPtr<CFG::Node>, CFG::stmt_iterator, CFG::succ_iterator>;
+	std::stack<NodeType> visitStack;
+	visitStack.push({node, stmtIter, node->succ_begin()});
```

Comment:
We could simplify
```cpp
visitStack.push({node, stmtIter, node->succ_begin()});
```
to
```cpp
visitStack.emplace(node, stmtIter, node->succ_begin());
```
(Emplacement allows us to get rid of the curly brackets.)